File: src/com/infoecos/cn/destiny/lib/TimezoneLocation.java
package com.infoecos.cn.destiny.lib;

import java.util.HashMap;
import java.util.Map;

public class TimezoneLocation {
	private double latitude = 1;
	private double longitude = 0;
	private float hoursOffset = 0;

	public static String[][] _timezone = { { "NZDT", "13f" },
			{ "IDLE", "12f" }, { "NZST", "12f" }, { "NZT", "12f" },
			{ "AESST", "11f" }, { "CST(ACSST)", "10.5f" }, { "CADT", "10.5f" },
			{ "SADT", "10.5f" }, { "EST(EAST)", "10f" }, { "GST", "10f" },
			{ "LIGT", "10f" }, { "CAST", "9.5f" }, { "SAT(SAST)", "9.5f" },
			{ "WDT(AWSST)", "9f" }, { "JST", "9f" }, { "KST", "9f" },
			{ "MT", "8.5f" }, { "WST(AWST)", "8f" }, { "CCT", "8f" },
			{ "JT", "7.5f" }, { "IT", "3.5f" }, { "BT", "3f" },
			{ "EETDST", "3f" }, { "CETDST", "2f" }, { "EET", "2f" },
			{ "FWT", "2f" }, { "IST", "2f" }, { "MEST", "2f" },
			{ "METDST", "2f" }, { "SST", "2f" }, { "BST", "1f" },
			{ "CET", "1f" }, { "DNT", "1f" }, { "FST", "1f" }, { "MET", "1f" },
			{ "MEWT", "1f" }, { "MEZ", "1f" }, { "NOR", "1f" },
			{ "SET", "1f" }, { "SWT", "1f" }, { "WETDST", "1f" },
			{ "GMT", "0f" }, { "WET", "0f" }, { "WAT", "-1f" },
			{ "NDT", "-2.5f" }, { "ADT", "-3f" }, { "NFT", "-3.5f" },
			{ "NST", "-3.5f" }, { "AST", "-4f" }, { "EDT", "-4f" },
			{ "CDT", "-5f" }, { "EST", "-5f" }, { "CST", "-6f" },
			{ "MDT", "-6f" }, { "MST", "-7f" }, { "PDT", "-7f" },
			{ "PST", "-8f" }, { "YDT", "-8f" }, { "HDT", "-9f" },
			{ "YST", "-9f" }, { "AHST", "-10f" }, { "CAT", "-10f" },
			{ "NT", "-11f" }, { "IDLW", "-12f" }, { "CCDT", "9f" }, };
	private static Map<String, Float> _timezoneMap = null;
	static {
		_timezoneMap = new HashMap<String, Float>();
		for (String[] tz : _timezone) {
			try {
				_timezoneMap.put(tz[0], Float.parseFloat(tz[1]));
			} catch (Exception e) {
			}
		}
	}

	public TimezoneLocation(double longitude, double latitude, String timezone) {
		this(timezone);

		this.latitude = latitude;
		this.longitude = longitude;
	}

	public TimezoneLocation(String timezone) {
		try {
			hoursOffset = Float.parseFloat(timezone);
		} catch (Exception e) {
			hoursOffset = 8f;
		}
		longitude = hoursOffset * 15;
	}

	public double getLatitude() {
		return latitude;
	}

	public double getLongitude() {
		return longitude;
	}

	public double getGMTAbsoluteHoursOffset() {
		return longitude / 15;
	}

	public double getGMTHoursOffset() {
		return hoursOffset;
	}
}

File: libs/destiny/TimezoneLocation.php
<?php

class TimezoneLocation extends Location {
	private $m_hoursOffset = 0;

	public static $_timezone = array(
		'NZDT' => 13, 
		'IDLE' => 12, 'NZST' => 12, 'NZT' => 12, 
		'AESST' => 11, 
		'CST(ACSST)' => 10.5, 'CADT' => 10.5, 'SADT' => 10.5, 'EST(EAST)' => 10, 'GST' => 10, 'LIGT' => 10, 
		'CAST' => 9.5, 'SAT(SAST)' => 9.5, 'WDT(AWSST)' => 9, 'JST' => 9, 'KST' => 9, 'CCDT' => 9,
		'MT' => 8.5, 'WST(AWST)' => 8, 'CCT' => 8, 
		'JT' => 7.5, 
		'IT' => 3.5, 'BT' => 3, 'EETDST' => 3, 
		'CETDST' => 2, 'EET' => 2, 'FWT' => 2, 'IST' => 2, 'MEST' => 2, 'METDST' => 2, 'SST' => 2, 
		'BST' => 1, 'CET' => 1, 'DNT' => 1, 'FST' => 1, 'MET' => 1, 'MEWT' => 1, 'MEZ' => 1, 'NOR' => 1, 'SET' => 1, 'SWT' => 1, 'WETDST' => 1, 
		'GMT' => 0, 'WET' => 0, 
		'WAT' => -1, 
		'NDT' => -2.5, 
		'ADT' => -3, 'NFT' => -3.5, 'NST' => -3.5, 
		'AST' => -4, 'EDT' => -4, 
		'CDT' => -5, 'EST' => -5, 
		'CST' => -6, 'MDT' => -6, 
		'MST' => -7, 'PDT' => -7, 
		'PST' => -8, 'YDT' => -8, 
		'HDT' => -9, 'YST' => -9, 
		'AHST' => -10, 'CAT' => -10, 
		'NT' => -11, 
		'IDLW' => -12, 
	);

	function __construct($timezone = 'CCT') {
		if(!isset(self::$_timezone[$timezone])) $timezone = 'CCT';
		$this->m_hoursOffset = self::$_timezone[$timezone];
		$this->m_longitude = $this->m_hoursOffset * 15;
	}

	public function getLatitude() {
		return $this->m_latitude;
	}

	public function getLongitude() {
		return $this->m_longitude;
	}

	public function getGMTAbsoluteHoursOffset() {
		return $this->m_longitude / 15;
	}

	public function getGMTHoursOffset() {
		return $this->m_hoursOffset;
	}
}
